src/Microsoft.AspNet.Mvc.Core/ActionContext.cs

```diff
@@ -56,23 +70,71 @@ public ActionContext([NotNull] ActionContext actionContext)
         }
 
         /// <summary>
-        /// Gets the <see cref="Mvc.ActionDescriptor"/> for the selected action.
+        /// Gets or sets the <see cref="Mvc.ActionDescriptor"/> for the selected action.
         /// </summary>
-        public ActionDescriptor ActionDescriptor { get; }
+        public ActionDescriptor ActionDescriptor
+        {
+            get
+            {
+                // This should run only for the controller unit test scenarios.
+                _actionDescriptor = _actionDescriptor ?? new ActionDescriptor();
+                return _actionDescriptor;
+            }
+            set
+            {
+                _actionDescriptor = value;
+            }
+        }
 
         /// <summary>
-        /// Gets the <see cref="Http.HttpContext"/> for the current request.
+        /// Gets or sets the <see cref="Http.HttpContext"/> for the current request.
         /// </summary>
-        public HttpContext HttpContext { get; }
+        public HttpContext HttpContext
+        {
+            get
+            {
+                // This should run only for the controller unit test scenarios.
+                _httpContext = _httpContext ?? new DefaultHttpContext();
+                return _httpContext;
```

Comment:
My understanding was that we just want to have a setter and document that you can use it for testing. I wouldn't pre-initialize any of these